### PR TITLE
GH-4328 FedX unclosed iteration

### PR DESCRIPTION
**Pull request by hmottestad**
GitHub issue resolved: #4328 

Briefly describe the changes proposed in this PR:

----
PR Author Checklist (see the [contributor guidelines](https://github.com/eclipse/rdf4j/blob/main/CONTRIBUTING.md) for more details):

 - [ ] my pull request is [self-contained](https://rdf4j.org/documentation/developer/merge-strategy/#self-contained-changes-pull-requests-and-commits)
 - [ ] I've added tests for the changes I made
 - [ ] I've applied [code formatting](https://github.com/eclipse/rdf4j/blob/main/CONTRIBUTING.md#code-formatting) (you can use `mvn process-resources` to format from the command line)
 - [ ] I've [squashed](https://rdf4j.org/documentation/developer/squashing) my commits where necessary 
 - [ ] every commit message starts with the issue number (GH-xxxx) followed by a meaningful description of the change

### CODE REVIEW

File: tools/federation/src/main/java/org/eclipse/rdf4j/federated/evaluation/SailTripleSource.java [hunk 1]
```diff
@@ -109,9 +112,11 @@ public CloseableIteration<Statement, QueryEvaluationException> getStatements(
 
 		return withConnection((conn, resultHolder) -> {
 
-			RepositoryResult<Statement> repoResult = conn.getStatements(subj, pred, obj,
-					queryInfo.getIncludeInferred(), contexts);
+			RepositoryResult<Statement> repoResult = null;
 			try {
+				repoResult = conn.getStatements(subj, pred, obj,
+						queryInfo.getIncludeInferred(), contexts);
```
Comment (hmottestad) [inline tools/federation/src/main/java/org/eclipse/rdf4j/federated/evaluation/SailTripleSource.java:118]: A few of the fixes are simply to make sure that we always close resources if there is an exception. So mostly moving the variables outside of the try-catch and adding a null check to the catch block.

File: tools/federation/src/main/java/org/eclipse/rdf4j/federated/evaluation/concurrent/ControlledWorkerScheduler.java [hunk 2]
```diff
@@ -201,32 +201,27 @@ public void toss(Exception e) {
...
 
-			ParallelExecutor<T> taskControl = task.getControl();
+			try {
 
-			if (taskControl.isFinished()) {
-				return;
-			}
+				if (aborted || Thread.currentThread().isInterrupted() || taskControl.isFinished()) {
+					throw new InterruptedException();
+				}
```
[the remark this answers is not in the document]
Comment (aschwarte10) [reply tools/federation/src/main/java/org/eclipse/rdf4j/federated/evaluation/concurrent/ControlledWorkerScheduler.java:223]: I would need to think and look a bit more what the "aborted" state actually means (i.e. if it is also kind of valid state where only partial results are requested from the client). In this case we would need to make sure that the exception is not preventing the consumption of the partial results.

Another question: why would you want to kill the threads entirely? They are controlled by a thread-pool and are re-used for operations. If I remember correctly the thread pool has some time-to-live before it removes threads.

Note: this is currently just out of my head - I do not have a chance to look closer into the code right now (could do in the beginning of next year after the Christmas / New Years time).

For the `isInterrupted()` case it is perfectly fine to propagate the exception

**Review by aschwarte10 — COMMENTED** [review]
Other than my comment on the `isAborted` (see above) the change looks good. Thanks for going through the code and resolving potential issues.

Please have a look at my comment / question above. For me it will only be possible to look at this again in the beginning of January. Feel free to go ahead, if the question above is resolved (and maybe someone else has a second look)

File: tools/federation/src/main/java/org/eclipse/rdf4j/federated/evaluation/concurrent/ControlledWorkerScheduler.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (hmottestad) [reply tools/federation/src/main/java/org/eclipse/rdf4j/federated/evaluation/concurrent/ControlledWorkerScheduler.java:223]: I don't know what `isFinished()` or `aborted` mean either. The problem that I see is that `ControlledWorkerUnion` uses a Phaser to wait for all the tasks to complete (I think), and that is why I think we should always call either `toss(...)` or `done()`.

I wonder if there are any existing solutions within Java or some kind of library that would provide a more robust and documented solution to solving these concurrency issues.

[the remark this answers is not in the document]
Comment (hmottestad) [reply tools/federation/src/main/java/org/eclipse/rdf4j/federated/evaluation/concurrent/ControlledWorkerScheduler.java:223]: Also, about the threads. What I'm seeing is that the test sometimes stalls for a while after it's actually done. It seems to be because it's waiting for the thread pools to terminate and at least one thread is stuck waiting on a phaser.

[the remark this answers is not in the document]
Comment (aschwarte10) [reply tools/federation/src/main/java/org/eclipse/rdf4j/federated/evaluation/concurrent/ControlledWorkerScheduler.java:223]: Ok, I had another look into the code and tried to remind myself about the implementation. I think it should be fine to throw the InterruptedException also in the `aborted` state: looks like this is some artifact which currently is not used at all. Not sure if it rather should be removed entirely.

What I meant above with consumption of partial results / cancelling queries, is done on another level (i.e. when scheduling the operations for intermediate results)

If you have any alternative to using the phaser for blocking until tasks are completed, I am happy to learn about it.

**Review by aschwarte10 — COMMENTED** [review]
Good idea to split the PR into two parts to get the more straight-forward changes already in 👍 

I think I am now also good with this change after having a second look. When I have initially worked on the concurrency a few years back, especially the concurrency unit test helped me a lot to get things right. Of course here it is more about the error / edge cases, but still I do have good trust in the unit tests. 

If you are also happy, go ahead and merge so that we can observe in the wild. I'll also keep an eye on our federation use-cases.
